components/avatar/avatar.vue

```diff
@@ -103,36 +94,108 @@ export default {
       type: Object,
       default: () => ({}),
     },
+
+    /**
+     * Initials to be shown in the avatar. Used as fallback if image fails to load.
+     */
+    initials: {
+      type: String,
+      default: '',
+    },
   },
 
   data () {
     return {
       // initials, image or icon
       kind: 'initials',
       AVATAR_SIZE_MODIFIERS,
-      AVATAR_COLOR_MODIFIERS,
       AVATAR_KIND_MODIFIERS,
       AVATAR_PRESENCE_SIZE_MODIFIERS,
+      imageLoadedSuccessfully: null,
+      formattedInitials: '',
     };
   },
 
+  computed: {
+    shouldShowContent () {
+      return this.kind !== 'initials' && this.imageLoadedSuccessfully !== false;
+    },
+
+    showInitials () {
+      return this.kind === 'initials' || this.initials;
+    },
+
+    initialKindStyle () {
+      const randomGradientColorStops = this.randomizeGradientColorStops();
+      const styles = {
+        '--avatar-gradient-angle': `${this.randomizeGradientAngle()}deg`,
+        '--avatar-gradient-stop-1': `var(--${randomGradientColorStops[0]})`,
+        '--avatar-gradient-stop-2': `var(--${randomGradientColorStops[1]})`,
+        '--avatar-gradient-stop-3': `var(--${randomGradientColorStops[2]})`,
+      };
+
+      return (this.kind === 'initials' || this.kind === 'icon') ? styles : null;
+    },
+  },
+
   mounted () {
     this.init();
+    this.validateImageAttrsPresence();
   },
 
   updated () {
     this.init();
   },
 
   methods: {
-    init () {
-      const firstChild = this.$el.firstChild;
-      if (firstChild) {
-        this.setKind(firstChild);
-        if (this.kind === 'image') {
-          firstChild.classList.add('d-avatar__image');
+    async init () {
+      const firstChild = this.$slots.default[0]?.elm || this.$slots.default[0];
+      const iconChild = this.$slots.default[0]?.componentOptions?.tag === 'dt-icon';
+
+      if (firstChild || iconChild) {
+        if (iconChild) {
+          this.kind = 'icon';
+        } else {
+          this.setKind(firstChild);
+        }
+
+        await this.$nextTick();
+        const childEl = this.$el;
+
+        if (this.kind === 'image' && this.imageLoadedSuccessfully === null) {
+          childEl.firstChild.classList.add('d-avatar__image');
+
+          childEl.firstChild.addEventListener('error', () => {
+            this.kind = 'initials';
+            this.imageLoadedSuccessfully = false;
+            childEl.firstChild?.classList?.remove('d-avatar__image');
+          });
+
+          childEl.firstChild.addEventListener('load', () => {
+            this.imageLoadedSuccessfully = true;
+            childEl.classList.add('d-avatar--image-loaded');
+          });
+
+          if (this.initials) {
+            this.formatInitials(this.initials);
+          }
+        }
+
+        if (this.kind === 'initials') {
+          this.formatInitials(firstChild.text || firstChild.textContent);
         }
```

Comment:
`formatInitials` is called twice inside the `init` method now. One in case the image fails to load, which means it'll fallback to initials and the other is just when the avatar is initial type.